# Colaborative filtering

Colaborative filtering is a type of recommender systems that makes recommendations (predictions) based only on collected ratings/preferences of many users. To quote Wikipedia:

> In the newer, narrower sense, collaborative filtering is a method of making automatic predictions (filtering) about the interests of a user by collecting preferences or taste information from many users (collaborating). The underlying assumption of the collaborative filtering approach is that if a person A has the same opinion as a person B on an issue, A is more likely to have B's opinion on a different issue than that of a randomly chosen person. 

### Rating matrix

The user ratings can be collected in a rating matrix

$$R_{um}$$

each row of this matrix corresponds to one user and each column to one item (movie in this example). 
 An entry $R_{um}$ gives the rating of user $u$ for item $m$. This matrix has size $N_u\times N_m$ where $N_u$ is the number of users and $N_m$ is the numbers of items. This can be a very large number. However ss each of the users has usually rated a (very) small sample of items this matrix is sparse! In case of e.g. Netflix it has only about $1\%$ of filled entries.  

The aim of the recommender system is to predict the remaining $99\%$ of entries :) 

### Matrix Factorisation

One approach to this problem is to reduce the  rank of the rating matrix by representing it as a product of two matrices:

$$\hat{R}_{um} = U_{uk} M_{mk},\quad \hat{R}=U\cdot M^T$$

In this formulation each entry of the rating matrix $\hat R$ is a dot product of a row of users matrix $U$ and a row of movies matrix $M$. We can view each row of movie matrix as a set of _features_ and each row of the users matrix as _preferences_ for each of the features.

The number of features $K$ is a hyperparameter that we have to choose somehow. 

Reconstructing the rating matrix $\hat R$ amounts to estimating the matrices $U$ and $M$. This is alltogether  $N_u\times K + N_u\times K$ parameters. Depending on $K$ this is usually much smaller number then the number of all entries in $\hat R$.  

Estimation of $U$ and $M$ is done byminimazing the mean squared error:

$$J=\frac{1}{2}\sum_{<u,m>}\left(R_{um}-\hat{R}_{um}\right)^2$$

The sum is over a set of all non empty entries of matrix $R$ and we denote this by symbol $\langle u,m\rangle$.

Differentiating with respect to an element of the matrix $U$ we obtain:

$$\frac{\partial }{\partial {U_{uk}}} J = \sum_{<u,m>} \left(R_{um} - U_{uk'} M_{mk'}\right) M_{mk}$$

In the above formula idices $u$ and $k$ are fixed and the sum runs over all the movies $m$ rated by the user $u$. 


Equating this derivative to zero gives as an equation

$$\sum_{<u,m>} M_{mk}  M_{mk'} U_{uk'} = \sum_{<u,m>}  R_{um} M_{mk}$$

which can be rewritten in vector form:

$$ \left(\sum_{<u,m>}\vec{M}_{m}  \vec{M}_{m}^T\right) \vec{U}_{u} = \sum_{<u,m>}  R_{um} \vec{M}_{m}$$

here $\vec{U}_u$ is the $u$-th row of the matrix $U$ and similarly for $\vec{M}_m$. The summation is again over all movies $m$ rated by user $u$. 

In the same way we can derive another set of equation for the rows of the matrix $M$:

$$ \left(\sum_{<u,m>}\vec{U}_{u}  \vec{U}_{u}^T\right) \vec{M}_{m} = \sum_{<u,m>}  R_{um} \vec{U}_{u}$$

Minimazing the  $J$  requires solving both of those sets of equations simultaneuosly. This cannot be done in general and requires numerical methods. One of the is the iterative merthods of alterneting least squares.

### Alternating least squares

I if we consider $M$ fixed the equiations for the rows of $U$ are ordinary linear equations. Based on this observation the method consists of filling $M$ with same random starting values and the solving for $U$

$$  \vec{U}_{u} = \left(\sum_{<u,m>}\vec{M}_{m}  \vec{M}_{m}^T\right)^{-1}\sum_{<u,m>}  R_{um} \vec{M}_{m}$$

Then keeping $U$ fixed we solve for $M$

$$ \vec{M}_{m} =\left(\sum_{<u,m>}\vec{U}_{u}  \vec{U}_{u}^T\right)  \sum_{<u,m>}  R_{um} \vec{U}_{u}$$

and the whole process is reapeted until convergence. 

## Data 

In [6]:
import numpy as np
import scipy as sp
import pandas as pd
from scipy.sparse import csr_matrix

We will use the data from the Movielens. Please download and unpack the  ```ml-latest-small.zip``` from their [website](https://grouplens.org/datasets/movielens/latest/).

In [7]:
data = pd.read_csv('ml-latest-small/ratings.csv')

In [8]:
data.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [9]:
data.drop('timestamp', axis=1, inplace=True)

In [10]:
n_ratings = len(data)
data.shape

(100836, 3)

To find the number of users and movies in the data set we can use function ```nunique```:

In [11]:
n_users  = data['userId'].nunique()
n_movies = data['movieId'].nunique()
print(n_users, n_movies)

610 9724


### Problem 1

Find the size of the rating matrix, assuming $K=64$ calculate number of parameters we have to fit. 

In [12]:
n_of_parameters = (n_users + n_movies) * 64
print(n_of_parameters)

661376


In [13]:
sparse_rating_matrix = csr_matrix((data.rating.values, (data.userId.values, data.movieId.values)))

In [14]:
sparse_rating_matrix.toarray()

array([[0. , 0. , 0. , ..., 0. , 0. , 0. ],
       [0. , 4. , 0. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ],
       ...,
       [0. , 2.5, 2. , ..., 0. , 0. , 0. ],
       [0. , 3. , 0. , ..., 0. , 0. , 0. ],
       [0. , 5. , 0. , ..., 0. , 0. , 0. ]])

Compare number of non empty entries in the rating matrix to number of parameters.

In [15]:
print("NON EMPTY ENTRIES: " + str(data.shape[0]))
print("num of parameters: " + str(n_of_parameters))

NON EMPTY ENTRIES: 100836
num of parameters: 661376


Compare number of paremeters to the size of available data. 

In [16]:
ratings_shape = sparse_rating_matrix.shape
print("Available data: " + str(ratings_shape[0] * ratings_shape[1]))
print("num of parameters: " + str(n_of_parameters))

Available data: 118295710
num of parameters: 661376


In [17]:
print('Number of movies that didnt appear in the dataset: ' + str(len(np.where(~sparse_rating_matrix.toarray().any(axis=0))[0]))) 

Number of movies that didnt appear in the dataset: 183886


#### Regularisation

Those calculations show that we still have more parameters to fit then the data. This can (and will) lead to serious overfitting. One way to mitigate that is to use regularisation of the parameters by constraining their size. 

This amounts to adding another term to the cost function $J$:

$$
J=\frac{1}{2}\sum_{<u,m>}\left(R_{um}-\hat{R}_{um}\right)^2+
\frac{\lambda}{2}\left(
||U||^2 +||M||^2
\right)
$$

where the norm of the matrix $||U||^2$ denotes the sum of all the squares of the matrix  entries

$$||U||^2=\sum_{u,k}U_{ik}^2$$

The $\lambda$ is another hyperparameter to be estimated.  

After taking the gradient we obtain new sets of equations

$$ \vec{U}_{u} = \left(\sum_{<u,m>}\vec{M}_{m}  
\vec{M}_{m}^T+\lambda I\right)^{-1}
\sum_{<u,m>} R_{um}\vec{M}_{m}$$

$$ \vec{M}_{m} =\left(\sum_{<u,m>}\vec{U}_{u}  \vec{U}_{u}^T+\lambda I\right)  \sum_{<u,m>}  R_{um} \vec{U}_{u}$$

that we solve in the same iterative way. 

Before proceeding to implementation we need to split our data into train, test and validation sets. 

In [18]:
train = data.sample(frac=0.8, replace=False)

In [19]:
train.head()

userId  movieId  rating
56350     372     1483     1.0
69948     448    86880     1.5
39655     274     2539     3.0
41061     278     3148     4.0
78854     489     4085     2.0

In [20]:
test = data.drop(train.index)

In [21]:
validation = test.sample(frac=0.5, replace=False)

In [22]:
validation.head()

userId  movieId  rating
7488       51     1088     4.0
12994      83     1862     0.5
40776     275     2357     4.0
67120     434      648     4.0
67241     434     5502     2.5

In [23]:
test = test.drop(validation.index)

In [24]:
test.head()

userId  movieId  rating
5        1       70     3.0
17       1      316     3.0
22       1      367     4.0
37       1      648     3.0
41       1      736     3.0

In [25]:
n_train_users  = train.userId.nunique()
n_train_movies = train.movieId.nunique()

In [26]:
print(n_train_users, n_train_movies)

610 8953


In [27]:
n_test_users  = test.userId.nunique()
n_test_movies = test.movieId.nunique()

In [28]:
print(n_test_users, n_test_movies)

601 3668


In [29]:
n_validation_users=validation.userId.nunique()
n_validation_movies = validation.movieId.nunique()

In [30]:
print(n_validation_users, n_validation_movies)

596 3703


When implementing the ALS algorithm we will need to traverse the ratings users by user and for each user iterate over all movies that she has rated. To this end we will define an auxiliary data structur

In [31]:
by_user = train.groupby('userId').apply(lambda g: list(zip(g.movieId, g.rating)))

This is a ```pandas.Series``` indexed by the  ```userId``` that contains lists of pairs ```(movieId, rating)``` for each movie rated by this partuclar user.

Similarly we will need to traverse list of all movies and for each movie traverse list of users that have rated it, so we create another auxiliary structure. 

In [32]:
by_movie = train.groupby('movieId').apply(lambda g: list(zip(g.userId, g.rating)))

One last point to remember is that we cannot(!) use the ```userId``` to index  rows of matrix $U$! This is because the userId are not garantied to be consecutive, especially after spliting the data into three different sets. The correct way of finding the index into the U matrix for given userId is to find it's location in the series:

```ui = by_user.index.get_loc(uid)```

Same applies to movies indexes

```mi = by_movie.index.get_loc(mid)```

In [33]:
test.head()

userId  movieId  rating
5        1       70     3.0
17       1      316     3.0
22       1      367     4.0
37       1      648     3.0
41       1      736     3.0

### Problem 2 

2.1 Implement the ALS algorithm in python. 

In [29]:
class AlternatingLeastSquares:
    def __init__(
        self,
        n_features: int = 64,
        n_iter: int = 30,
        λ: float = 4.44,
        𝜀: float = 0.1
    ):
        self.n_features = n_features
        self.n_iter = n_iter
        self.λ = λ
        self.𝜀 = 𝜀
        self.features_eye = np.eye(n_features)

    def mean_absolute_error(self, y_pred, y_true):
        nonZero = y_true != 0
        return np.mean(abs(y_pred[nonZero] - y_true[nonZero]))
    
    def mean_squared_error(self, y_pred, y_true):
        nonZero = y_true != 0
        return np.mean(np.square(y_pred[nonZero] - y_true[nonZero]))

    def normaliseRow(self, x):
        return x / sum(x)

    def generate_matrix(self, shape):
        matrix = abs(np.random.rand(shape[0], shape[1]))
        return np.apply_along_axis(self.normaliseRow, 1, matrix)
    
    def fit(self, X):
        self.U = self.generate_matrix((X.shape[0], self.n_features))
        self.M = self.generate_matrix((X.shape[1], self.n_features))
        nonZero = X > 0
        
        for i in range(self.n_iter):
            print('Iteration', i + 1, 'of', self.n_iter)
            for i in range(X.shape[0]):
                idx = nonZero[i,:]
                self.U[i] = self._factorise(X[i,idx], self.M[idx,], self.λ, self.features_eye)

            for i in range(X.shape[1]):
                idx = nonZero[:,i]
                self.M[i] = self._factorise(X[idx,i], self.U[idx,], self.λ, self.features_eye)
        
            mse = self.mean_squared_error(self.predict(), X)
            print('Mean Squared Error:', mse)
            if mse < self.𝜀:
                print('Fitting finished, epsilon =', self.𝜀,'bound reached with MSE =', mse)
                return
            
    def predict(self):
        return self._predict(self.U, self.M)
    
    def _predict(self, U, M):
        return U.dot(M.T)
            
    def _factorise(self, r, X, λ, eye):
        a = np.dot(X.T, X) + (λ * eye)
        b = np.dot(r.T, X)
        return np.linalg.solve(a, b)

2.2 Check its predictions on the train, test and validation set. 

In [30]:
model = AlternatingLeastSquares()

mae = model.mean_absolute_error
mse = model.mean_squared_error

In [31]:
ratings_train = train.pivot_table(columns=['movieId'], index=['userId'], values='rating', dropna=False)
ratings_train = ratings_train.fillna(0).values
print(ratings_train.shape)
ratings_train

(610, 8939)


array([[ 4. ,  0. ,  4. , ...,  0. ,  0. ,  0. ],
       [ 0. ,  0. ,  0. , ...,  0. ,  0. ,  0. ],
       [ 0. ,  0. ,  0. , ...,  0. ,  0. ,  0. ],
       ..., 
       [ 2.5,  2. ,  2. , ...,  0. ,  0. ,  0. ],
       [ 3. ,  0. ,  0. , ...,  0. ,  0. ,  0. ],
       [ 5. ,  0. ,  0. , ...,  0. ,  0. ,  0. ]])

In [32]:
model.fit(ratings_train)

Iteration 1 of 30
Mean Squared Error: 2.33150221665
Iteration 2 of 30
Mean Squared Error: 0.161975072226
Iteration 3 of 30
Mean Squared Error: 0.121959164792
Iteration 4 of 30
Mean Squared Error: 0.112802210327
Iteration 5 of 30
Mean Squared Error: 0.108988804768
Iteration 6 of 30
Mean Squared Error: 0.106862948836
Iteration 7 of 30
Mean Squared Error: 0.105487005159
Iteration 8 of 30
Mean Squared Error: 0.104523235091
Iteration 9 of 30
Mean Squared Error: 0.103815216799
Iteration 10 of 30
Mean Squared Error: 0.103276579449
Iteration 11 of 30
Mean Squared Error: 0.102854621869
Iteration 12 of 30
Mean Squared Error: 0.10251498397
Iteration 13 of 30
Mean Squared Error: 0.102234471126
Iteration 14 of 30
Mean Squared Error: 0.101997378534
Iteration 15 of 30
Mean Squared Error: 0.101793164395
Iteration 16 of 30
Mean Squared Error: 0.101614731456
Iteration 17 of 30
Mean Squared Error: 0.101457168006
Iteration 18 of 30
Mean Squared Error: 0.101316902639
Iteration 19 of 30
Mean Squared Error: 

In [33]:
predictions = model.predict()
print(predictions.shape)
predictions

(610, 8939)


array([[ 3.92216972,  3.29647753,  3.9223446 , ...,  1.61267886,
         1.61267886,  2.3919555 ],
       [ 3.79872412,  3.1964166 ,  3.02416462, ...,  1.26072015,
         1.26072015,  1.61126753],
       [ 1.88499399,  2.03764836,  1.24621029, ...,  0.94225207,
         0.94225207,  0.98876234],
       ..., 
       [ 2.58985167,  2.1367159 ,  2.14880352, ...,  1.74056803,
         1.74056803,  2.05995896],
       [ 3.16743808,  2.91079989,  2.58092212, ...,  0.88677819,
         0.88677819,  1.24715716],
       [ 5.07054208,  3.64392468,  5.11079251, ...,  2.58975748,
         2.58975748,  2.49875894]])

One problem  is that the validation and test sets may contain users and/or movies that are not present in the training set. In this case the prediction is impossible. We can check this at the level of caclulating the prediction score or clean up those sets beforehand.  We will clean up the sets using an inner join on two dataframes.  

First we prepare two dataframes that contain only index which constists of unique users(movies) Id from the training sets. 

In [34]:
by_user_idx = pd.DataFrame(index = by_user.index)
by_movie_idx = pd.DataFrame(index = by_movie.index)

In [35]:
len(test)

10083

We now use this frames in join. Inner join keeps only the rows with keys apearing in both dataframes. 

In [36]:
test = test.join(by_user_idx, on='userId', how='inner')
test = test.join(by_movie_idx, on='movieId', how='inner')

In [37]:
test.head()

userId  movieId  rating
17          1      316     3.0
3908       24      316     3.5
5615       40      316     3.0
14175      91      316     4.0
41772     284      316     4.0

In [38]:
validation = validation.join(by_user_idx, on='userId', how='inner')
validation = validation.join(by_movie_idx, on='movieId', how='inner')

In [39]:
validation.head()

userId  movieId  rating
95404     600     2081     4.0
49132     318     2081     3.0
6675       45     2081     5.0
57152     380     2081     4.0
31686     219     2081     3.0

In [40]:
def locate(a, b):
    ids_set = set(a)
    idx = []
    for i, entry in enumerate(b):
        if entry in ids_set:
            idx.append(i)
            ids_set.remove(entry)
    return (np.array(idx), np.array(list(ids_set)))

In [41]:
test_users_idx, test_users_idx_notfound = locate(test.userId.unique(), train.userId.unique())
test_movies_idx, test_movies_idx_notfound = locate(test.movieId.unique(), train.movieId.unique())

print('users found:', len(test_users_idx))
print('users not found:', len(test_users_idx_notfound))
print('movies found:', len(test_movies_idx))
print('movies not found:', len(test_movies_idx_notfound))

users found: 596
users not found: 0
movies found: 3286
movies not found: 0


In [42]:
predictions_test = predictions[test_users_idx]
predictions_test = predictions_test[:,test_movies_idx]
print(predictions_test.shape)
predictions_test

(596, 3286)


array([[ 3.92216972,  3.29647753,  3.9223446 , ...,  1.17553849,
         0.94728329,  1.46214781],
       [ 3.79872412,  3.1964166 ,  3.02416462, ...,  0.7220557 ,
         0.48262781,  1.14710354],
       [ 1.88499399,  2.03764836,  1.24621029, ...,  0.60196544,
         0.3665341 ,  0.43201557],
       ..., 
       [ 2.58985167,  2.1367159 ,  2.14880352, ...,  1.41877319,
         0.81400686,  1.25945483],
       [ 3.16743808,  2.91079989,  2.58092212, ...,  0.48376213,
         0.3330671 ,  0.92868206],
       [ 5.07054208,  3.64392468,  5.11079251, ...,  2.12376305,
         1.84470522,  1.29207233]])

In [43]:
ratings_test = test.pivot_table(columns=['movieId'], index=['userId'], values='rating', dropna=False)
ratings_test = ratings_test.fillna(0).values
print(ratings_test.shape)
ratings_test

(596, 3286)


array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])

In [44]:
validation_users_idx, validation_users_idx_notfound = locate(validation.userId.unique(), train.userId.unique())
validation_movies_idx, validation_movies_idx_notfound = locate(validation.movieId.unique(), train.movieId.unique())

print('users found:', len(validation_users_idx))
print('users not found:', len(validation_users_idx_notfound))
print('movies found:', len(validation_movies_idx))
print('movies not found:', len(validation_movies_idx_notfound))

users found: 604
users not found: 0
movies found: 3254
movies not found: 0


In [45]:
predictions_validation = predictions[validation_users_idx]
predictions_validation = predictions_validation[:,validation_movies_idx]
print(predictions_validation.shape)
predictions_validation

(604, 3254)


array([[ 3.92216972,  3.29647753,  3.9223446 , ...,  0.54299925,
         1.03435142,  1.15035616],
       [ 3.79872412,  3.1964166 ,  3.02416462, ...,  0.3581282 ,
         0.76077139,  0.8236973 ],
       [ 1.88499399,  2.03764836,  1.24621029, ...,  0.36509465,
         0.54416901,  0.52409704],
       ..., 
       [ 2.58985167,  2.1367159 ,  2.14880352, ...,  0.5636208 ,
         1.16815333,  0.83081188],
       [ 3.16743808,  2.91079989,  2.58092212, ...,  0.24224505,
         0.56304109,  0.49075274],
       [ 5.07054208,  3.64392468,  5.11079251, ...,  1.52123848,
         1.26927771,  1.87968247]])

In [46]:
ratings_validation = validation.pivot_table(columns=['movieId'], index=['userId'], values='rating', dropna=False)
ratings_validation = ratings_validation.fillna(0).values
print(ratings_validation.shape)
ratings_validation

(604, 3254)


array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ..., 
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.]])

In [47]:
print('-----------------------')
print('Model parameters:')
print('    K -', model.n_features)
print('    λ -', model.λ)
print('    max_i -', model.n_iter)
print('-----------------------')
print('[Training set] MSE:', mse(predictions, ratings_train))
print('[Training set] MAE:', mae(predictions, ratings_train))
print('[Test set] MSE:', mse(predictions_test, ratings_test))
print('[Test set] MAE:', mae(predictions_test, ratings_test))
print('[Validation set] MSE:', mse(predictions_validation, ratings_validation))
print('[Validation set] MAE:', mae(predictions_validation, ratings_validation))

-----------------------
Model parameters:
    K - 64
    λ - 4.44
    max_i - 30
-----------------------
[Training set] MSE: 0.100345612976
[Training set] MAE: 0.22042818263
[Test set] MSE: 4.78289479915
[Test set] MAE: 1.84838821859
[Validation set] MSE: 4.25433221786
[Validation set] MAE: 1.72141525047


2.3 Does the result depend on number of features (K)? How?

Large K definitely hits the performance (time complexity) 

2.4 Does the result depend on $\lambda$? How?

With lower $\lambda$ model tends to overfit the predictions

### Biases

We can extend our model by adding additional "biases" and mean:

$$\hat{R}_{um} = U_{uk} M_{mk} + b_u +c_m +\mu$$

where $\mu$ is just overall mean (on training set):

$$\mu = \frac{1}{\sum_{<u,m> } 1}\sum_{<u,m>}R_{um}$$

We can calculate gradient with respwect to biases

$$\frac{\partial }{\partial {b_{u}}} J = \sum_{<u,m>} \left(R_{um} - U_{uk'} M_{mk'}- b_u -c_m -\mu\right) \cdot 1 $$

$$\sum_{<u,m>} \left(R_{um} - U_{uk'} M_{mk'}- b_u -c_m -\mu\right)  $$

$$\sum_{<u,m>}  b_u =\sum_{<u,m>} \left(R_{um} - U_{uk'} M_{mk'} -c_m -\mu\right)  $$

and obtain te final equations:

$$ \vec{U}_{u} = \left(\sum_{<u,m>}\vec{M}_{m}  
\vec{M}_{m}^T+\lambda I\right)^{-1}
\sum_{<u,m>} \left( R_{um} - b_u -c_m -\mu\right)\vec{M}_{m}$$

$$ \vec{M}_{m} =\left(\sum_{<u,m>}\vec{U}_{u}  \vec{U}_{u}^T+\lambda I\right)  \sum_{<u,m>} \left( R_{um} - b_u -c_m -\mu\right) \vec{U}_{u}$$

$$ b_u =\frac{1}{\lambda+\sum_{<u,m>} 1}\sum_{<u,m>} \left(R_{um} - U_{uk'} M_{mk'} -c_m -\mu\right)  $$

$$ c_m =\frac{1}{\lambda+\sum_{<u,m>} 1}\sum_{<u,m>} \left(R_{um} - U_{uk'} M_{mk'} -b_u -\mu\right)  $$

We solve this equations in the same way, we start from random values and  solve each equation by keeping all other varaibles fix. 

### Problem 3

3.1 Implement this version of the ALS algorithm and compare it's performance with the previous one.

In [87]:
class ALSB:
    def __init__(
        self,
        n_features: int = 64,
        n_iter: int = 1,
        λ: float = 4.44,
        𝜀: float = 0.1
    ):
        self.n_features = n_features
        self.n_iter = n_iter
        self.λ = λ
        self.𝜀 = 𝜀
        self.features_eye = np.eye(n_features)

    def mean_absolute_error(self, y_pred, y_true):
        nonZero = y_true != 0
        return np.mean(abs(y_pred[nonZero] - y_true[nonZero]))
    
    def mean_squared_error(self, y_pred, y_true):
        nonZero = y_true != 0
        return np.mean(np.square(y_pred[nonZero] - y_true[nonZero]))

    def normaliseRow(self, x):
        return x / sum(x)

    def generate_matrix(self, shape):
        matrix = abs(np.random.rand(shape[0], shape[1]))
        return np.apply_along_axis(self.normaliseRow, 1, matrix)
    
    def fit(self, X, mu):
        self.U = self.generate_matrix((X.shape[0], self.n_features))
        self.M = self.generate_matrix((X.shape[1], self.n_features))
        self.Ub = np.zeros(X.shape[0])
        self.Mb = np.zeros(X.shape[1])
        self.Gb = mu
        nonZero = X > 0
        
        for i in range(self.n_iter):
            print('Iteration', i + 1, 'of', self.n_iter)
            #Update biases
            for i in range(X.shape[0]):
                idxU = nonZero[i,:]
                self.Ub[i] = self._bias(X[i,idxU], self.U[i], self.M[idxU,], self.λ, self.Mb[idxU,], self.Gb)
            break
            for i in range(X.shape[1]):
                idxM = nonZero[:,i]
                self.Mb[i] = self._bias(X[idxM,i], self.U[idxM,], self.M[i], self.λ, self.Ub[idxM,], self.Gb)
            
            for i in range(X.shape[0]):
                idx = nonZero[i,:]
                self.U[i] = self._factorise(X[i,idx], self.M[idx,], self.λ, self.features_eye, Ub[i], Mb[idx,], self.Gb)

            for i in range(X.shape[1]):
                idx = nonZero[:,i]
                self.M[i] = self._factorise(X[idx,i], self.U[idx,], self.λ, self.features_eye, Ub[idx,], Mb[i], self.Gb)
        
            mse = self.mean_squared_error(self.predict(), X)
            print('Mean Squared Error:', mse)
            if mse < self.𝜀:
                print('Fitting finished, epsilon =', self.𝜀,'bound reached with MSE =', mse)
                return
            
    def predict(self):
        return self._predict(self.U, self.M)
    
    def _predict(self, U, M):
        return U.dot(M.T)
    
    def _bias(self, r, U, M, λ, b, Gb):
        print( r.shape, U.shape, M.shape, b.shape)
        a = (r - np.dot(U,M) - b - Gb)
        b = (len(r) + λ)
        return a/b
            
    def _factorise(self, r, X, λ, eye, Ub, Mb, Gb):
        a = np.dot(X.T, X) + (λ * eye)
        b = np.dot((r.T-Ub-Mb-Gb), X)
        return np.linalg.solve(a, b)

In [88]:
model = ALSB()

#mae = model.mean_absolute_error
#mse = model.mean_squared_error

In [89]:
mu = np.mean( train[['rating']] )

ratings_train = train.pivot_table(columns=['movieId'], index=['userId'], values='rating', dropna=False)
ratings_train = ratings_train.fillna(0).values
print(ratings_train.shape)
ratings_train

(610, 8953)


array([[4. , 0. , 4. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ],
       ...,
       [2.5, 0. , 2. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 0. ]])

In [90]:
model.fit(ratings_train, mu)

Iteration 1 of 1
(176,) (64,) (176, 64) (176,)


ValueError: shapes (64,) and (176,64) not aligned: 64 (dim 0) != 176 (dim 0)

3.2 Using the ```movies.csv``` set associate the movies with biases. Print top ten movies with highest bias, and top ten with lowest (negative). 